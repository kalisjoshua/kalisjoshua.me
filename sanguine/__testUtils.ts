import { createFileNode } from "./createFileNode.ts";

const ROOT = "/path/to/projects/pro";

const createTestFileNode = (rel: string, raw: string) => ({
  ...createFileNode(ROOT, `${ROOT}/content/${rel}`),
  raw,
});

export { createTestFileNode, ROOT };
