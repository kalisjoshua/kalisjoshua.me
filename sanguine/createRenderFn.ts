import { Link } from "./createNavigation.ts";
import { createNavigation } from "./createNavigation.ts";

interface PageContext {
  isActive?: (h: string) => string;
  isArticle?: boolean;
  main: string;
  navigation: Array<Link>;
  path?: string;
  rel: string;
  section: string;
}

const rConditionalBlock = /\{\{\?(.*?)\}\}([\w\W]*?)\{\{\?\/}\}/im;

function resolveBlocks(context: PageContext, str: string) {
  let result = str;
  let match;

  while ((match = result.match(rConditionalBlock))) {
    const key = match[1] as keyof PageContext;

    result = result.replace(match[0], context[key] ? match[2].trim() : "");
  }

  return result;
}

function createRenderFn({
  meta,
  template,
  ...rest
}: ReturnType<typeof createNavigation>) {
  const fill = new Function(
    "context",
    `with(context){return \`${template.raw}\`}`,
  );

  return {
    ...rest,
    render: (pageContext: PageContext) => {
      const fullContext: PageContext = { ...meta, ...pageContext };

      return resolveBlocks(fullContext, fill(fullContext));
    },
  };
}

export type { PageContext };
export { createRenderFn };
