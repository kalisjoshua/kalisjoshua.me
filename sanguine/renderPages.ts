import { marked } from "../deno.deps.ts";

import { FileNodeContent } from "./addFileNodeContent.ts";
import { createRenderFn } from "./createRenderFn.ts";

type SiteContent = ReturnType<typeof createRenderFn>;

function genPage(siteContent: SiteContent, page: FileNodeContent) {
  const { navigation, render } = siteContent;
  const path = `${page.section && `${page.section}/`}${page.name}.html`;

  return {
    html: render({
      isActive: (href: string) =>
        href === path || (page.section && href.startsWith(page.section))
          ? "active"
          : "",
      isArticle: page.section === "articles" && page.name !== "index",
      main: marked.parse(page.raw),
      navigation,
      path,
      rel: page.rel,
      section: page.section || page.name,
    }),
    path,
  };
}

function renderPages(siteContent: SiteContent) {
  return siteContent.pages.map((page) => genPage(siteContent, page));
}

export { renderPages };
