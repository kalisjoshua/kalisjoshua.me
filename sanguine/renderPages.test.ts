import { assertSnapshot } from "../deno.deps.ts";

import { createTestFileNode } from "./__testUtils.ts";
import { createRenderFn, PageContext } from "./createRenderFn.ts";

import { renderPages } from "./renderPages.ts";

type SiteContent = ReturnType<typeof createRenderFn>;

const isActive = (href: string, section: string, path = "") =>
  href === path || (!!section && href.startsWith(section)) ? "isActive" : "";
const render = ({ main, navigation, path, rel, section }: PageContext) => `
<ul>
${navigation
  .map(
    ({ href, text }) =>
      `<li><a class="${isActive(
        href,
        section,
        path
      )}" href="${rel}${href}">${text}</a></li>`
  )
  .join("\n")}
</ul>
<main>
${main}
</main>
`;
const siteContent: SiteContent = {
  pages: [
    createTestFileNode("pages/about.md", "# Hello\n\n\nAbout me."),
    createTestFileNode("pages/index.md", "# Resume"),

    createTestFileNode("pages/articles/index.md", "# Articles Index"),
    createTestFileNode("pages/articles/something.md", "# Somthing"),
    createTestFileNode("pages/articles/another.md", "# Another"),
  ],
  navigation: [
    { href: "index.html", text: "Resume" },
    { href: "about.html", text: "About" },
    { href: "articles/index.html", text: "Articles" },
  ],
  render,
};

Deno.test("renderPages", async (testContext) => {
  const result = renderPages(siteContent).flatMap(
    (page: Record<string, string>) => `${page.path}\n${page.html}`.split("\n")
  );

  // this formatting - array of strings (one per line) - is preferrable to the
  // default multiline formatting done by Deno
  await assertSnapshot<string[]>(testContext, result);
});
