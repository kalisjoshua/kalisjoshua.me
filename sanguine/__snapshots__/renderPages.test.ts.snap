export const snapshot = {};

snapshot[`renderPages 1`] = `
[
  "about.html",
  "",
  "<ul>",
  '<li><a class="" href="./index.html">Resume</a></li>',
  '<li><a class="isActive" href="./about.html">About</a></li>',
  '<li><a class="" href="./articles/index.html">Articles</a></li>',
  "</ul>",
  "<main>",
  '<h1 id="hello">Hello</h1>',
  "<p>About me.</p>",
  "",
  "</main>",
  "",
  "index.html",
  "",
  "<ul>",
  '<li><a class="isActive" href="./index.html">Resume</a></li>',
  '<li><a class="" href="./about.html">About</a></li>',
  '<li><a class="" href="./articles/index.html">Articles</a></li>',
  "</ul>",
  "<main>",
  '<h1 id="resume">Resume</h1>',
  "",
  "</main>",
  "",
  "articles/index.html",
  "",
  "<ul>",
  '<li><a class="" href="../index.html">Resume</a></li>',
  '<li><a class="" href="../about.html">About</a></li>',
  '<li><a class="isActive" href="../articles/index.html">Articles</a></li>',
  "</ul>",
  "<main>",
  '<h1 id="articles-index">Articles Index</h1>',
  "",
  "</main>",
  "",
  "articles/something.html",
  "",
  "<ul>",
  '<li><a class="" href="../index.html">Resume</a></li>',
  '<li><a class="" href="../about.html">About</a></li>',
  '<li><a class="isActive" href="../articles/index.html">Articles</a></li>',
  "</ul>",
  "<main>",
  '<h1 id="somthing">Somthing</h1>',
  "",
  "</main>",
  "",
  "articles/another.html",
  "",
  "<ul>",
  '<li><a class="" href="../index.html">Resume</a></li>',
  '<li><a class="" href="../about.html">About</a></li>',
  '<li><a class="isActive" href="../articles/index.html">Articles</a></li>',
  "</ul>",
  "<main>",
  '<h1 id="another">Another</h1>',
  "",
  "</main>",
  "",
]
`;
